src/librbd/cache/pwl/AbstractWriteLog.cc

```diff
@@ -1551,11 +1542,21 @@ bool AbstractWriteLog<I>::check_allocation(
       m_bytes_allocated += bytes_allocated;
       m_bytes_cached += bytes_cached;
       m_bytes_dirty += bytes_dirtied;
+      if (m_cache_state->clean && m_bytes_dirty) {
+        m_cache_state->clean = false;
+        update_image_cache_state();
+        need_update_state = true;
+      }
     } else {
       alloc_succeeds = false;
     }
   }
 
+  if (need_update_state) {
+    std::unique_lock locker(m_lock);
+    write_image_cache_state(locker);
```

Comment:
Is `need_update_state` variable still needed?  There is no loop involved and `m_lock` is held above.  I think you could call `write_image_cache_state` right after `update_image_cache_state`.

Reply:
Yes, needed.  `update_image_cache_state` is under `lock_guard`, but `write_image_cache_state` should be under `unique_lock`. Because the `m_lock` needs to be unlocked at the end of `write_image_cache_state` call stack due to https://github.com/ceph/ceph/pull/47325.

Reply:
I probably understand what you mean. Then remove the flag `need_update_state` and add `unlock()` in the `else` branch. Is that so?

Reply:
Almost -- I don't think a manual unlock and therefore any kind of `else` branch is needed.